client/components/swipeable/index.js

```diff
@@ -0,0 +1,249 @@
+import classnames from 'classnames';
+import { useRtl } from 'i18n-calypso';
+import { Children, useState, useEffect, useRef } from 'react';
+
+import './style.scss';
+
+export const Swipeable = ( {
+	hasDynamicHeight = false,
+	children,
+	currentPage = 0,
+	onPageSelect,
+	pageClassName,
+	containerClassName,
+} ) => {
+	const [ pageWidth, setPageWidth ] = useState( null );
+	const [ swipeableArea, setSwipeableArea ] = useState();
+	const isRtl = useRtl();
+
+	const [ pagesStyle, setPagesStyle ] = useState( {
+		transform: `translate3d(0px, 0px, 0px)`,
+		transitionDuration: `300ms`,
+	} );
+	const [ dragStartData, setDragStartData ] = useState();
+
+	const pagesRef = useRef();
+	const resizeObserverRef = useRef();
+	const numPages = Children.count( children );
+
+	const getWidth = () => {
+		return resizeObserverRef.current?.getBoundingClientRect().width;
+	};
+
+	const getPagesWidth = () => {
+		return pageWidth * numPages;
+	};
+
+	const getOffset = ( index ) => {
+		const offset = pageWidth * index;
+		return isRtl ? offset : -offset;
+	};
+
+	function useUpdateLayout( enabled, currentPageIndex, updateLayout ) {
```

Comment:
I'm a bit confused by the `useUpdateLayout` function, maybe @tyxla can also help out in case I'm missing something:

- it looks like this function is supposed to be used like a React hook, but it is declared inside the render function of `Swipeable` — should it not be defined outside this component?
- I'm not sure I get why we need to use a `ref` to store/invoke the `updateLayout` callback — wouldn't it be easier to wrap that callback in a `useCallbak` hook where it is defined, instead of assigning it to a ref? It makes this code much harder to follow

Reply:
> it looks like this function is supposed to be used like a React hook, but it is declared inside the render function of Swipeable — should it not be defined outside this component?

I agree, it should be defined outside of the component.

> I'm not sure I get why we need to use a ref to store/invoke the updateLayout callback — wouldn't it be easier to wrap that callback in a useCallbak hook where it is defined, instead of assigning it to a ref? It makes this code much harder to follow

I think what Marco's suggesting makes sense. This likely came to be like that because of an earlier suggestion by @jsnajdr in another PR.

Reply:
yes this part of the code is there because of https://github.com/Automattic/wp-calypso/pull/58377

Reply:
> I'm not sure I get why we need to use a ref to store/invoke the updateLayout callback — wouldn't it be easier to wrap that callback in a useCallbak hook where it is defined, instead of assigning it to a ref? It makes this code much harder to follow

Regarding this comment, this is what I mean:

```js
const updateLayoutCallback = useCallback(
  ( hasTransitionDuration = true ) => { /* function body */ },
  [ /* dependencies */ ]
);

useUpdateLayout( updateEnabled, currentPage, updateLayoutCallback );
```

Since now the callback is memoized, you could simplify `useUpdatedLayout` and remove the `ref` — something like

```js
function useUpdateLayout( enabled, currentPageIndex, updateLayout ) {
	// fire when the `currentPageIndex` parameter changes
	useEffect( () => {
		if ( ! enabled ) {
			return;
		}

		updateLayout();
	}, [ enabled, currentPageIndex, updateLayout ] );

	// fire when the window resizes
	useEffect( () => {
		if ( ! enabled ) {
			return;
		}

		const onResize = () => updateLayout( false );
		window.addEventListener( 'resize', onResize );
		window.addEventListener( 'orientationchange', onResize );
		return () => {
			window.removeEventListener( 'resize', onResize );
			window.removeEventListener( 'orientationchange', onResize );
		};
	}, [ enabled, updateLayout ] );
}
```

I haven't tested it on my machine, but I hope it makes sense!